### PR TITLE
Bring belochub/jstp-cli upstream

### PR DESCRIPTION
**Pull request by aqrln**
I've created a [branch](https://github.com/metarhia/JSTP/tree/cli) to port [belochub/jstp-cli](https://github.com/belochub/jstp-cli/) to be a part of JSTP itself, as it was agreed upon in https://github.com/belochub/jstp-cli/issues/1.

Right now `cli.js` is located in the root of the project in order to mirror the original repo and make cherry-picking new commits effortless. It should later be separated into `bin/cli.js` and `lib/cli/*.js`.

/cc @belochub @lundibundi

### CODE REVIEW

**Comment by aqrln** [comment]
@lundibundi I've grabbed the latest patches.

File: tools/cli.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+rl.on('line', (line) => {
+  const [type, leftover] = _split(line.trim(), ' ', 1);
+  if (!type) {
+    return rl.prompt(true);
+  }
+  const processor = lineProcessor[type];
+  if (!processor) {
+    log(`Unknown command ${type}`);
+  } else {
```
Comment (aqrln) [inline tools/cli.js]: I'd prefer either the if branch to return from the function and reduce the indentation level for the code inside the else branch or to flip the condition. I don't feel very strongly for one way or other, but I like the first one more.

File: tools/cli.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    processor(leftover, (err, result) => {
+      if (err) return log(`${err.name} occurred: ${err.message}`);
```
Comment (aqrln) [inline tools/cli.js]: I'd find it more readable to maybe use braces here.

File: tools/cli.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  return [[], line];
+}
+
+const state = {
+  client: null,
+  connection: null
+};
+
+commandProcessor.call = (interfaceName, methodName, args, callback) => {
+  if (state.client === null) return callback(new Error('Not connected'));
```
Comment (aqrln) [inline tools/cli.js]: `if (!state.client)` would be more readable here.

Comment (aqrln) [inline tools/cli.js]: Not sure about whether the braces would help readability since the next line is quite long too.

File: tools/cli.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  state.connection.callMethod(interfaceName, methodName, args, callback);
+};
+
+commandProcessor.event = (interfaceName, eventName, args, callback) => {
+  if (state.client === null) return callback(new Error('Not connected'));
```
Comment (aqrln) [inline tools/cli.js]: Ditto.

File: tools/cli.js [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  state.client.connectAndHandshake(appName, null, null,
+      (err, connection) => {
+        if (err) return callback(err);
+        state.connection = connection;
+        // todo make event registering generic
+        connection.on('event', (data) => {
+          log(`Received remote event: ${jstp.stringify(data)}`);
+        });
+        callback();
+      });
```
Comment (aqrln) [inline tools/cli.js]: I'd prefer to put `);` on a new line.

File: tools/cli.js [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+commandProcessor.event = (interfaceName, eventName, args, callback) => {
+  if (state.client === null) return callback(new Error('Not connected'));
+  state.connection.emitRemoteEvent(interfaceName, eventName, args);
+  callback();
+};
+
+commandProcessor.connect = (host, port, appName, callback) => {
+  state.client = jstp.tcp.createClient({ host, port, secure: true });
+  state.client.connectAndHandshake(appName, null, null,
```
Comment (aqrln) [inline tools/cli.js]: And maybe `appName, null, null` too.

File: tools/cli.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        // todo make event registering generic
+        connection.on('event', (data) => {
+          log(`Received remote event: ${jstp.stringify(data)}`);
+        });
+        callback();
+      });
+};
+
+commandProcessor.disconnect = (callback) => {
+  if (state.client !== null) {
```
Comment (aqrln) [inline tools/cli.js]: `if (state.client)` would be sufficient.

File: tools/cli.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+  process.exit();
+};
+
+function _split(str, separator, limit, leaveEmpty) {
+  const shouldTrim = (start, split) => !leaveEmpty && start === split;
+
+  const result = [];
+  let start = 0;
+
+  for (let i = 0; !limit || i < limit; i++) {
```
Comment (aqrln) [inline tools/cli.js]: Is it intentional that zero limit is no limit?

File: tools/cli.js [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+  callback();
+};
+
+commandProcessor.connect = (host, port, appName, callback) => {
+  state.client = jstp.tcp.createClient({ host, port, secure: true });
+  state.client.connectAndHandshake(appName, null, null,
+      (err, connection) => {
+        if (err) return callback(err);
+        state.connection = connection;
+        // todo make event registering generic
```
Comment (aqrln) [inline tools/cli.js]: The proper spelling of `todo` is `TODO:` :) That'll make it highlighted in code editors and picked up to be shown on the bitHound dashboard.

File: tools/cli.js [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+};
+
+lineProcessor.disconnect = (_, callback) => {
+  commandProcessor.disconnect((err) => {
+    if (err) return callback('Not connected');
+    callback(null, 'Successful disconnect');
+  });
+};
+
+// map all remaining commands directly
```
Comment (aqrln) [inline tools/cli.js]: Can you please start the comment with a capital letter?

File: tools/cli.js (hunk 1 above)
Comment (belochub) [inline tools/cli.js]: The problem here is that using return is not the same as using else.

File: tools/cli.js [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+const commandProcessor = {};
+const lineProcessor = {};
+
+const rl = readline.createInterface({
+  input: process.stdin,
+  output: process.stdout,
+  completer
+});
+
+const log = msg => {
```
Comment (lundibundi) [inline tools/cli.js]: not a problem, should we add it to eslint ([rule](http://eslint.org/docs/rules/arrow-parens))?

File: tools/cli.js (hunk 3 above)
Comment (lundibundi) [inline tools/cli.js]: @aqrln ok with ```!state.client```, but the braces will not help that much

File: tools/cli.js (hunk 5 above)
Comment (lundibundi) [inline tools/cli.js]: sure

File: tools/cli.js (hunk 6 above)
Comment (lundibundi) [inline tools/cli.js]: @aqrln not sure about this one, then wouldn't it be better to put each argument on separate line, but I see no reason for that here. @belochub what's your opinion?

File: tools/cli.js (hunk 7 above)
Comment (lundibundi) [inline tools/cli.js]: sure

File: tools/cli.js (hunk 6 above)
Comment (belochub) [inline tools/cli.js]: It is okay either way, though I'd prefer to leave it as it is now.

File: tools/cli.js (hunk 8 above)
Comment (lundibundi) [inline tools/cli.js]: yes

File: tools/cli.js (hunk 9 above)
Comment (lundibundi) [inline tools/cli.js]: oh, okay

File: tools/cli.js (hunk 10 above)
Comment (lundibundi) [inline tools/cli.js]: eh, [meme](https://www.mememaker.net/static/images/memes/4486420.jpg), okay

File: tools/cli.js (hunk 11 above)
Comment (aqrln) [inline tools/cli.js]: This rule with `as-needed` and `requireForBlockBody` options enabled is exactly what we use. Would you like to open a PR for that?

Comment (lundibundi) [inline tools/cli.js]: ok

**Comment by aqrln** [comment]
@lundibundi for the record: commit your amendments based on the feedback from the review to this branch.

**Comment by aqrln** [comment]
@belochub and @lundibundi, thank you a lot, you've done a great job shaping it up. I'm going to land this so that new improvements can be done incrementally as separate PRs, but it needs reviewers to sign-off the commits. All your commits LGTM for me, but I need you to review my de10f00da36ec43a38ad017b21d5734c39d7c61b and I'd also like to know whether you have reviewed each other's commits.

**Comment by lundibundi** [comment]
LGTM, it's good to be able to start it as usual shell command. Also yeah we have reviewed each other's commits so it's okay to land.

**Comment by belochub** [comment]
@aqrln, you forgot to add https://github.com/belochub/jstp-cli/commit/5ecc3ca46989f1976637be60888ab812c767fde1

**Comment by aqrln** [comment]
@belochub just did it ;) Thanks!

**Comment by aqrln** [comment]
@lundibundi 

> cli: fix 'disconnect' error handling bug
> Author: lundibundi <lundibundi@users.noreply.github.com>
> Date: Fri May 5 00:18:28 2017 +0300

Did you do that from the GitHub UI?

**Comment by lundibundi** [comment]
@aqrln yeah, it's correct. I wanted to test it out, I thought github can fetch my email address.

**Comment by aqrln** [comment]
Oh, gosh, Travis has restarted.

**Comment by aqrln** [comment]
Landed in 4e298cbb21e1815c28f730c6dbc272a243994cfc...6fba474d3d493e65372503bb85c7c3a68ef54220.

@belochub and @lundibundi thanks for working on this! 🎉

**Pull request closed without merging**